engine/classes/Elgg/Router.php

```diff
@@ -70,42 +74,50 @@ public function route(\Elgg\Http\Request $request) {
 			$this->timer->begin(['build page']);
 		}
 
+		ob_start();
+
 		$result = $this->hooks->trigger('route', $identifier, $old, $old);
 		if ($result === false) {
-			return true;
-		}
-
-		if ($result !== $old) {
+			$output = ob_get_clean();
+			$response = elgg_ok_response($output);
+		} else if ($result !== $old) {
 			_elgg_services()->logger->warn('Use the route:rewrite hook to modify routes.');
-		}
+			$query = parse_url(current_page_url(), PHP_URL_QUERY);
+			$segments = (array) elgg_extract('segments', $result, []);
+			$identifier = elgg_extract('identifier', $result);
+			array_unshift($segments, $identifier);
+			_elgg_services()->responseFactory->redirect(implode('/', $segments) . '?' . $query);
+			exit;
+		} else {
 
-		if ($identifier != $result['identifier']) {
-			$identifier = $result['identifier'];
-		} else if ($identifier != $result['handler']) {
-			$identifier = $result['handler'];
-		}
+			if ($identifier != $result['identifier']) {
+				$identifier = $result['identifier'];
+			} else if ($identifier != $result['handler']) {
+				$identifier = $result['handler'];
+			}
 
-		$segments = $result['segments'];
+			$segments = $result['segments'];
 
-		$handled = false;
-		ob_start();
+			$response = false;
 
-		if (isset($this->handlers[$identifier]) && is_callable($this->handlers[$identifier])) {
-			$function = $this->handlers[$identifier];
-			$handled = call_user_func($function, $segments, $identifier);
-		}
+			if (isset($this->handlers[$identifier]) && is_callable($this->handlers[$identifier])) {
+				$function = $this->handlers[$identifier];
+				$response = call_user_func($function, $segments, $identifier);
+			}
 
-		$output = ob_get_clean();
+			$output = ob_get_clean();
 
-		$ajax_api = _elgg_services()->ajax;
-		if ($ajax_api->isReady()) {
-			$path = implode('/', $request->getUrlSegments());
-			$ajax_api->respondFromOutput($output, "path:$path");
-			return true;
+			if ($response === false) {
+				return headers_sent();
+			}
+
+			if (!$response instanceof ResponseBuilder) {
+				$response = elgg_ok_response($output);
+			}
 		}
 
-		echo $output;
-		return $handled || headers_sent();
+		_elgg_services()->responseFactory->respond($response);
+		return true;
```

Comment:
The workaround is to keep a reference of the sent response and read through `ResponseFactory::getSentResponse`